### PR TITLE
Added speech synthesis bypass engine

### LINKED ISSUE #2435: Microphone button disabled when speech synthesis is disabled and received an "expecting" input hint
## Screenshots

image

## Version

`master`

## Describe the bug

When speech synthesis is disabled, but speech recognition is enabled (i.e. sample 06.h), the microphone button is disabled after receiving an "expecting" input hint.

## To Reproduce

1. Run sample `06.h`
2. Say, "input hint expected"

## Expected behavior

After receiving the message "This activity is expecting input", the microphone should be re-opened to receive further speech commands from the user.

## Additional context

Looks like the expecting input is trying to open the microphone, but somehow it stuck at "opening (1)" phase.

image

When we are investigating, the "expecting input" saga is trying to turn on the microphone by calling `startDictate`. This will open microphone *after* all synthesis is completed, which never happen. But if we call `stopSpeakingActivity`, it will never speak any activities if speech synthesis is not disabled.

Looks like the best fix would be, instead of ignoring the speech synthesis engine, we should provide a "no-op" speech synthesis engine.

If the dev did not provide a speech synthesis engine, the "no-op" engine will pretend to synthesize the activity and signal the completion synchronously. In this way, the expecting input should continue to work.

[Bug]

### PR DESCRIPTION
**Pull request by compulim**
> Fixes #2435.

### Fixed

-  Fix [#2435](https://github.com/microsoft/BotFramework-WebChat/issues/2435). Fix microphone button stuck on voice-triggered expecting input hint without a speech synthesis engine, by [@compulim](https://github.com/compulim) in PR [#2445](https://github.com/microsoft/BotFramework-WebChat/pull/2445)

## Description

On a system without speech synthesis engine:

- Queue a message to be synthesized
- Open microphone in a programmatic way and not clicking on microphone button

Since speech synthesis is not available, the queued message is pending to be synthesized and never complete. Internally, we will wait until all queued messages are synthesized before turning on the microphone. This is to prevent echo (bot hearing itself talking). Thus, the microphone will get stuck and never turned on.

The speech synthesis engine is configured on the `component` side. When the saga in the `core` side tries to perform synthesis, it has no knowledge whether the engine is available or not.

This can be reproduced by:

- Configure off speech synthesis
- Click microphone button and trigger the bot to respond with expecting input hint
- Click microphone button again

Since no speech synthesis engine is configured, the bot message is being queued but never complete. And the microphone is queued to open after all synthesis is completed. Thus, the microphone button is disabled and never open.

## Specific Changes

If speech synthesis engine is not configured, we will provide a bypass engine.

The bypass engine will act just like the real engine, except all synthesis operation will be completed instantly. This will unblock any future issues related to queued messages on a configured off system.

In the future, we could consider moving all speech operations to `component` side to prevent this issue from happening.

---

-  [x] Testing Added

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -96,6 +96,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 -  Fix [#2360](https://github.com/microsoft/BotFramework-WebChat/issues/2360). Timestamp should update on language change, by [@compulim](https://github.com/compulim) in PR [#2414](https://github.com/microsoft/BotFramework-WebChat/pull/2414)
 -  Fix [#2428](https://github.com/microsoft/BotFramework-WebChat/issues/2428). Should interrupt speech synthesis after microphone button is clicked, by [@compulim](https://github.com/compulim) in PR [#2429](https://github.com/microsoft/BotFramework-WebChat/pull/2429)
 -  Fix [#2422](https://github.com/microsoft/BotFramework-WebChat/issues/2422). Store thumbnail URL using the activity's `attachment.thumbnailUrl` field, by [@compulim](https://github.com/compulim) in PR [#2433](https://github.com/microsoft/BotFramework-WebChat/pull/2433)
+-  Fix [#2435](https://github.com/microsoft/BotFramework-WebChat/issues/2435). Fix microphone button stuck on voice-triggered expecting input hint without a speech synthesis engine, by [@compulim](https://github.com/compulim) in PR [#2445](https://github.com/microsoft/BotFramework-WebChat/pull/2445)
```
Comment (tdurnford) [inline CHANGELOG.md]:
```suggestion
-  Fix [#2435](https://github.com/microsoft/BotFramework-WebChat/issues/2435). Fix microphone button getting stuck on voice-triggered expecting input hint without a speech synthesis engine, by [@compulim](https://github.com/compulim) in PR [#2445](https://github.com/microsoft/BotFramework-WebChat/pull/2445)
```

File: packages/component/src/Speech/BypassSpeechSynthesisPonyfill.js [hunk 2]
```diff
@@ -0,0 +1,163 @@
+// Since this is a bypass, we will relax some ESLint rules
+
+/* eslint class-methods-use-this: "off" */
+/* eslint getter-return: "off" */
+/* eslint max-classes-per-file: ["error", 4] */
+/* eslint no-empty-function: "off" */
+
```
Comment (corinagum) [inline packages/component/src/Speech/BypassSpeechSynthesisPonyfill.js]: leave a comment: these are all the props defined in w3c

File: packages/component/src/external/event-target-shim.js [hunk 3]
```diff
@@ -0,0 +1,836 @@
...
+ * @typedef {object} PrivateData
+ * @property {EventTarget} eventTarget The event target.
+ * @property {{type:string}} event The original event object.
+ * @property {number} eventPhase The current event phase.
+ * @property {EventTarget|null} currentTarget The current event target.
+ * @property {boolean} canceled The flag to prevent default.
+ * @property {boolean} stopped The flag to stop propagation.
+ * @property {boolean} immediateStopped The flag to stop propagation immediately.
+ * @property {Function|null} passiveListener The listener if the current listener is passive. Otherwise this is null.
+ * @property {number} timeStamp The unix time.
```
[the remark this answers is not in the document]
Comment (compulim) [reply packages/component/src/external/event-target-shim.js]: Not touching this file.

File: packages/component/src/external/event-target-shim.js [hunk 4]
```diff
@@ -0,0 +1,836 @@
+// This is adopted from event-target-shim@5.0.1 under MIT License.
+// The source code is copied here because the original package do not support ES5 browsers.
+
+/* eslint-disable */
+
```
[the remark this answers is not in the document]
Comment (compulim) [reply packages/component/src/external/event-target-shim.js]: It's MIT licensed. It is okay to merge/copy the source code as long as the license is in-place.

**Pull request merged**